HPCC-19116 Error using DATASET(CSV(ESCAPE('')))

AttilaVamos opened the pull request:
Prevent to use empty string as Escape for CSV file

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->
Tested manually with the provided (in JIRA) ECL code

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

AttilaVamos commented:
@jakesmith please review.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: cc115aaa04238281347b8f2e67f202630f59d1dc
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |83 |83 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |765 |765 |0 |
|test (thor) |679 |679 |0 |
|test (roxie) |792 |792 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   14 sec (00:00:14) |  185 sec (00:03:05) |   55 sec (00:00:55) |    6 sec (00:00:06) |   37 sec (00:00:37) | 1140 sec (00:19:00) |   32 sec (00:00:32) | 1469 sec (00:24:29) |

richardkchapman commented on common/thorhelper/csvsplitter.cpp:
```diff
@@ -73,7 +73,8 @@ void CSVSplitter::addTerminator(const char * text)
 void CSVSplitter::addEscape(const char * text)
 {
-    matcher.addEntry(text, ESCAPE);
+    if (text && *text)
```
Is this the right fix? Seems like the user is specifically trying to set there to be "no escape character" - will this do that?

AttilaVamos replied:
If there was not defined "escape" at Spray or when CSV file created , then "No escape" (NULL) is the default. 
He received 'System error: -1: Duplicate entry "" added to string matcher' because he tried to add "no escape" (NULL) to the StringMatcher and it is throws an exception.  

The aim of my fix is to prevent this situation.

richardkchapman replied:
Sounds like the correct fix is to ignore duplicates ...

AttilaVamos commented on common/thorhelper/csvsplitter.cpp:
```diff
@@ -73,7 +73,7 @@ void CSVSplitter::addTerminator(const char * text)
 void CSVSplitter::addEscape(const char * text)
 {
-    matcher.addEntry(text, ESCAPE);
+    matcher.queryAddEntry((size32_t)strlen(text), text, ESCAPE);
```
@richardkchapman as you wish.

BTW: I think this is a bit over-sophisticated solution (add extra function call) to do a simple thing: if 'no escape' then do nothing.
Based on:
```
bool StringMatcher::queryAddEntry(unsigned len, const char * text, unsigned action)
{
    if (len == 0)
        return false;
```

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: e8fde246018ec8190ce948b48eac5e9d7f3039ee
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |83 |83 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |765 |765 |0 |
|test (thor) |679 |679 |0 |
|test (roxie) |792 |792 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   13 sec (00:00:13) |  182 sec (00:03:02) |   56 sec (00:00:56) |    6 sec (00:00:06) |   39 sec (00:00:39) | 1146 sec (00:19:06) |   32 sec (00:00:32) | 1474 sec (00:24:34) |

richardkchapman commented:
@ghalliday Can you sanity-check

ghalliday commented:
@AttilaVamos looks ok, please can you squash.

AttilaVamos committed:
HPCC-19116 Error using DATASET(CSV(ESCAPE('')))

Prevent to use empty string as Escape for CSV file

Tested with the provided (in JIRA) ECL code

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 850877e7d6d85a3fa4029d1fd0e419226b74ad7e
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |83 |83 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |759 |759 |0 |
|test (thor) |686 |686 |0 |
|test (roxie) |799 |799 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    7 sec (00:00:07) |  180 sec (00:03:00) |   55 sec (00:00:55) |    6 sec (00:00:06) |   21 sec (00:00:21) | 1141 sec (00:19:01) |   16 sec (00:00:16) | 1426 sec (00:23:46) |
